Add extensible background build discarders

daniel-beck opened the pull request:
(Description updated 2019-12-15)

This is the extensible implementation of #4336, allowing to pull out regex matching into a plugin.

A new extension point for configuring _global_ build discarders is added to core. These are run:

1. After a build finishes, for the given job.
2. Periodically.

It provides two simple implementations of a new extension point in core:

* Run the already configured per-project build discarder (can probably be considered a bug fix, and is therefore enabled by default).
* Run a specific, globally configured build discarder on all jobs.

Additionally, the API for the new extension point `BackgroundBuildDiscarderStrategy` should be flexible enough to support regexes and other complex criteria as discussed in the original issue #4336.

**Screenshot:**

Configuration UI:

<img width="1245" alt="Screenshot" src="https://user-images.githubusercontent.com/1831569/70868973-54fa7400-1f86-11ea-86bc-b81ef6d006a6.png">

**Manual testing hint:** Trigger periodic execution manually using `ExtensionList.lookupSingleton(BackgroundBuildDiscarder.class).doRun()` in script console.

**On classification:** _Major bug_ if the lack of periodic execution of the configured build rotator is considered a bug at all, as it could result in data loss with carelessly configured projects relying on the previous behavior.

**TODO List**:

- [x] Internalization
- [x] Basic test coverage
- [x] Help files
- [x] Better symbols ("global" is overloaded)
- [x] Rename classes and update labels as needed since it's no longer just "periodic"

**Open Questions** (feedback welcome):

- **Should the "Job" Global Build Discarder be implemented differently, without a UI option?** If we actually consider it to be a bug fix (see above and below), why would there be a UI that boils down to "[x] fix the bug"?
- **Should we really limit each descriptor to one occurrence?** This should make it more difficult to implement more flexible strategies, e.g. the proposed "regex match" one would need its own list of (regex/build discarder) pairs, rather than just one of each as a property, and then have N instances of the extension in the top level list…

### Proposed changelog entries

* Major RFE: Add globally configured build discarders that delete old builds not marked as "keep forever" even if there is no, or a less aggressive, per-project build discarder configured, executed periodically and after a build finishes.
* Major Bug: Jenkins will by default execute the configured per-project build discarder periodically even if no build is currently finishing. **This may delete old builds of projects that got a more aggressive build discarder configuration since the last build was run.**
* RFE: Developer: Add new extension point `BackgroundBuildDiscarderStrategy` to allow more flexible build discarding strategies for the global build discarder configuration.

### Submitter checklist

- [ ] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@awittha (author of #4336)
@res0nance (reviewer of #4336)

daniel-beck committed:
Add extensible background build discarders

Default implementation: Run per-job configured discarder periodically

awittha commented on core/src/main/java/jenkins/model/BackgroundBuildDiscarder.java:
```diff
@@ -0,0 +1,77 @@
+                        strategy.apply(job);
+                    } catch (Exception ex) {
+                        listener.error("An exception occurred when executing " + displayName + ": " + ex.getMessage());
+                        LOGGER.log(Level.WARNING, "An exception occurred when executing " + displayName, ex);
```
💯 for logging with listener _and_ the Java logger 👍

awittha commented on core/src/main/java/jenkins/model/BackgroundBuildDiscarderStrategy.java:
```diff
@@ -0,0 +1,82 @@
+            try {
+                apply(run);
+            } catch (IOException|InterruptedException ex) {
+                // TODO should these actually be caught, or just thrown up to stop applying?
```
IMO, `IOExceptoin` should be logged well, but chomped. One "bad" build in a project with 1000 builds shouldn't block attempts to clean up the other 999.

`InterruptedException` is a little less-certain: Since this would be happening _in the background_, what can raise an `InterruptedException` in a background task? If there's anything that thinks it will stop a BG task by interrupting it, we'd like to honor that semantic. 

If it's more of a "well, threads get interrupted sometimes and that checked exception is part of the Java interface," then, chomp & log along with `IOException`.

awittha commented on core/src/main/java/jenkins/model/GlobalBuildDiscarderStrategy.java:
```diff
@@ -0,0 +1,71 @@
+ */
+public class GlobalBuildDiscarderStrategy extends BackgroundBuildDiscarderStrategy {
+
+    private BuildDiscarder discarder;
```
Public class, all methods public, class neither `final` nor  marked `@NoExternalUse`, _but_... the key instance variable is `private`? 

I know there's a culture of `always private` around a lot of the Jenkins codebase, but this seems like an odd one out among the rest of an otherwise-extensible & reusable class.

daniel-beck replied:
Good point. It wasn't written with subclasses in mind; there's little benefit to doing that anyway, given the ratio of necessary new code (`Descriptor`!) to reusable code. I'm fairly sure I've experienced hierarchies in Jenkins actively hostile to subclasses due to how the generics worked out, plus there's always backward compatibility concerns because of plugins written a decade ago, and the problem of missing UI support for new supertype features… best not to start here if there's a perfectly good `ExtensionPoint` to work with. So I would not expect that this is something you'd extend in a plugin; rather, start with the abstract class directly. That's the most stable interface for extension in plugins, and no changes to this type (a future `[x] Skip jobs with a configured build discarder` perhaps?) could affect your subtype. There's just a handful of additional lines of duplication.

awittha replied:
> It wasn't written with subclasses in mind;

IMO, all non-`final` classes are implicitly written with subclasses in mind, because it's Java. The actively-hostile-to-subclasses parts of Jenkins feel so alien compared to the rest of my Java experience where extension is ... part of how one builds in the language.

Perhaps at least add a JavaDoc comment to the (abridged) effect of your above comment here "Don't extend this to make your own strageies; extend the abstract `BackgroundBuildDiscarderStrategy`"

daniel-beck replied:
Probably just going to make it `@Restricted` to prevent any use from plugins; cannot think of anything that would be expected besides configuration-as-code, and that ignores class-level `@Restricted`.

res0nance commented on core/src/main/resources/hudson/model/Run/logKeep.jelly:
```diff
@@ -27,7 +27,7 @@ THE SOFTWARE.
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
-  <j:if test="${it.parent.buildDiscarder!=null and it.canToggleLogKeep()}">
+  <j:if test="${it.canToggleLogKeep()}">
```
🥇 for remembering to enable `Keep Build Forever` button in the case where a explicit build discarder isn't set for the job

res0nance commented on core/src/main/java/jenkins/model/GlobalBuildDiscarderStrategy.java:
```diff
@@ -0,0 +1,71 @@
+        this.discarder = discarder;
+    }
+
+    public BuildDiscarder getDiscarder() {
```
Nit: `@CheckForNull` to catch potential bugs

timja commented:
I've tested this with jcasc, it works, but it's not possible to remove build discarders :(

i.e. this:
```
unclassified:
  globalBuildDiscarders:
    configuredBuildDiscarders:
    - globalBuildDiscarder:
        discarder:
          logRotator:
            artifactDaysToKeepStr: "15"
            artifactNumToKeepStr: "12"
            daysToKeepStr: "11"
            numToKeepStr: "10"
    - "jobBuildDiscarder"
```
to
```
unclassified:
  globalBuildDiscarders:
    configuredBuildDiscarders:
    - globalBuildDiscarder:
        discarder:
          logRotator:
            artifactDaysToKeepStr: "15"
            artifactNumToKeepStr: "12"
            daysToKeepStr: "11"
            numToKeepStr: "10"
```

doesn't remove `jobBuildDiscarder`

daniel-beck commented:
@timja Interesting. Did you try to change some of the values to see whether any kind of update works at all? Or what happens when you go from only `globalBuildDiscarder` to only `jobBuildDiscarder`?

daniel-beck commented:
Note to self: Come up with some symbols that don't result in
```
  globalBuildDiscarders:
    configuredBuildDiscarders:
    - globalBuildDiscarder:
```

timja commented:
> @timja Interesting. Did you try to change some of the values to see whether any kind of update works at all? Or what happens when you go from only `globalBuildDiscarder` to only `jobBuildDiscarder`?

I tried changing values and that worked

timja commented:
> @timja Interesting. Did you try to change some of the values to see whether any kind of update works at all? Or what happens when you go from only `globalBuildDiscarder` to only `jobBuildDiscarder`?

removing was ignored

daniel-beck commented:
@timja Thanks for that. Weird.

I basically use the same approach as `@Symbol(value="artifactManager")` does, to get databinding from `f:repeatableHeteroProperty` for free. Does the same problem exist there as well (hinting at a problem with core or JCasC rather than this PR)? Implementations there are from https://jenkins.io/doc/developer/extensions/jenkins-core/#artifactmanagerfactory

daniel-beck committed:
Prevent use of the specific strategy from plugins

timja commented:
> @timja Thanks for that. Weird.
> 
> I basically use the same approach as `@Symbol(value="artifactManager")` does, to get databinding from `f:repeatableHeteroProperty` for free. Does the same problem exist there as well (hinting at a problem with core or JCasC rather than this PR)? Implementations there are from [jenkins.io/doc/developer/extensions/jenkins-core/#artifactmanagerfactory](https://jenkins.io/doc/developer/extensions/jenkins-core/#artifactmanagerfactory)

No the same problem does not exist there,
I installed all 3 plugins,

Configured with this:
```
unclassified:
  artifactManager:
    artifactManagerFactories:
    - "compressing"
    - azure:
        config:
          container: "abcd"
          prefix: "abc"
```

And then I updated  it to:
```
unclassified:
  artifactManager:
    artifactManagerFactories:
    - azure:
        config:
          container: "abcdf"
          prefix: "abc"

```

It removed `compressing` and  updated the azure container config correctly

timja commented:
Weird I just tested this PR again and JCasC wise everything is working, (except some symbols would be nice to shorten the names)

daniel-beck commented:
@timja 

> some symbols would be nice to shorten the names

Which symbols are missing/wrong? Based on https://github.com/jenkinsci/jenkins/pull/4368#issuecomment-557782830 it looks OK to me. The ambiguous double use of "global" I already know and plan to fix.

timja commented:
> @timja
> 
> > some symbols would be nice to shorten the names
> 
> Which symbols are missing/wrong? Based on [#4368 (comment)](https://github.com/jenkinsci/jenkins/pull/4368#issuecomment-557782830) it looks OK to me. The ambiguous double use of "global" I already know and plan to fix.

original:
```
unclassified:
  globalBuildDiscarders:
    configuredBuildDiscarders:
    - globalBuildDiscarder:
        discarder:
          logRotator:
            artifactDaysToKeepStr: "15"
            artifactNumToKeepStr: "12"
            daysToKeepStr: "11"
            numToKeepStr: "10"
    - "jobBuildDiscarder"
```

possibly something like this (if symbols don't clash):
```
unclassified:
  globalBuildDiscarders:
    configured:
    - global:
        discarder:
          logRotator:
            artifactDaysToKeepStr: "15"
            artifactNumToKeepStr: "12"
            daysToKeepStr: "11"
            numToKeepStr: "10"
    - "job"
```

Ideally it would be flatted to not need the `discarder` and `configuredBuildDiscarders` or `configured` renamed to something else

but it works so 🤷‍♂

daniel-beck committed:
Apply global discarders after build finalization, i18n, etc.

daniel-beck commented:
This update now applies the globally configured build discarders after a build finishes so they're no longer only running once an hour. Labels are not yet updated accordingly.

daniel-beck commented:
> Ideally it would be flattened to not need the `discarder` and `configuredBuildDiscarders`

@timja 
I don't think the former is possible at all. The latter would require the extension class itself to be a `List`, which doesn't look like a great approach. Or how else would I do this?

timja commented:
> > Ideally it would be flattened to not need the `discarder` and `configuredBuildDiscarders`
> 
> @timja
> I don't think the former is possible at all. The latter would require the extension class itself to be a `List`, which doesn't look like a great approach. Or how else would I do this?

Hmm, been taking a look around,
I can't see a way, it looks like it just comes down to naming to make it sensible.

This is what the configuration now looks like:
```yaml
unclassified:
  buildDiscarders:
    configuredBuildDiscarders:
      - "jobBuildDiscarder"
      - globalBuildDiscarder:
          discarder:
            logRotator:
              daysToKeepStr: "7"
              numToKeepStr: "10"
```

timja commented on core/src/main/java/jenkins/model/GlobalBuildDiscarderConfiguration.java:
```diff
@@ -0,0 +1,72 @@
+            configuredBuildDiscarders.rebuildHetero(req, json, BackgroundBuildDiscarderStrategyDescriptor.all(), "configuredBuildDiscarders");
+            return true;
+        } catch (IOException x) {
+            throw new FormException(x, "artifactManagerFactories");
```
:shipit: 
```suggestion
            throw new FormException(x, "configuredBuildDiscarders");
```

daniel-beck commented on core/src/main/resources/jenkins/model/JobBackgroundBuildDiscarderStrategy/config.properties:
```diff
@@ -0,0 +1,3 @@
+blurb = By default, build discarders are only run after a build finishes. \
+  This option runs jobs' configured build discarders periodically, applying configuration changes even when no new builds are run. \
```
```suggestion
  This option runs jobs'' configured build discarders periodically, applying configuration changes even when no new builds are run. \
```

daniel-beck commented:
Considering this ready for review now. The initial PR comment tracks the TODO list, and I think it's little enough that this can be reviewed now.

daniel-beck committed:
Fix field name reference in string

daniel-beck committed:
Fix localized string

daniel-beck committed:
Rename the strategy

daniel-beck committed:
Wording changes in global build discarder UI

daniel-beck committed:
Stop using 'background' in favor of 'global'

(except for the periodic work thread)

daniel-beck committed:
Add missing license header

daniel-beck committed:
Fix compile error

daniel-beck committed:
This option no effect

daniel-beck commented:
I no longer consider this to be a work in progress, I've done everything here I set out to do.

timja commented on test/src/test/java/jenkins/model/GlobalBuildDiscarderTest.java:
```diff
@@ -0,0 +1,109 @@
+
+            // apply global config changes periodically
+            ExtensionList.lookupSingleton(BackgroundGlobalBuildDiscarder.class).execute(TaskListener.NULL);
+            Assert.assertArrayEquals("2 builds again", p.getBuilds().stream().mapToInt(Run::getNumber).toArray(), new int[]{10});
```
Could save a few characters for the reader and make it look a bit nicer if you static imported Assert.assertArrayEquals throughout this file

timja reviewed (approved):
Looks great

timja commented:
Requested review from everyone who reviewed the previous PR https://github.com/jenkinsci/jenkins/pull/4336

res0nance commented on core/src/main/java/jenkins/model/BackgroundGlobalBuildDiscarder.java:
```diff
@@ -0,0 +1,80 @@
+            String displayName = strategy.getDescriptor().getDisplayName();
+            listener.getLogger().println("Offering " + job.getFullName() + " to " + displayName);
+            if (strategy.isApplicable(job)) {
+                listener.getLogger().println(job.getFullName() + " accepted by " + displayName);
```
I'm slightly concerned that this might be excessive logging on very large instances

daniel-beck replied:
Maybe, but it's all going to a rotated set of log files, so will never fill up a disk (except when it's tiny to begin with):

https://github.com/jenkinsci/jenkins/blob/eedf2a4e2a25f5883d828f22c2429ceebe2bd157/core/src/main/java/hudson/model/AsyncPeriodicWork.java#L134-L156

res0nance reviewed (approved):
Overall, this looks fine to me

daniel-beck commented:
Added another _Open Question_ to the PR comment for reviewers' consideration.

jvz commented on core/src/main/java/jenkins/model/GlobalBuildDiscarderStrategy.java:
```diff
@@ -0,0 +1,82 @@
+            try {
+                apply(run);
+            } catch (IOException|InterruptedException ex) {
+                // TODO should these actually be caught, or just thrown up to stop applying?
```
Whenever you catch an `InterruptedException`, it's usually a good idea to re-set the interrupt flag again via `Thread.currentThread().interrupt();` Otherwise, you can end up with hung threads if they continue running (which in this case, it very well could).

daniel-beck replied:
So I would not actually want to catch the `InterruptedException` here? The idea was that if there's something wrong with deleting one build's artifacts, e.g. it taking too long to delete a billion files, an initial abort would continue with the next build, rather than just stop this process entirely.

Obviously the downside of that is that it's fairly difficult to really interrupt this thread while it's iterating here.

awittha replied:
> fairly difficult to really interrupt this thread while it's iterating here.

What processes will intentionally attempt such an interrupt? Is there any jenkins-initiated or user-initiated workflow that can target a specific AsyncPeriodicWork and try to stop it?

jvz replied:
Setting the interrupt flag doesn't re-throw the interrupted exception btw.

jvz replied:
I was about to leave a comment about this and noticed that I already did.

timja replied:
@daniel-beck Are you planning on making this change? either not catching the InterruptedException or setting the interrupt flag?

awittha commented on core/src/main/resources/jenkins/model/JobGlobalBuildDiscarderStrategy/config.properties:
```diff
@@ -0,0 +1,3 @@
+blurb = Build discarders configured for a job are only run after a build finishes. \
+  This option runs jobs'' configured build discarders periodically, applying configuration changes even when no new builds are run. \
```
> applying configuration changes even when no new builds are run.

This wording seems a bit confusing to me. From where will configuration changes that can be applied to the job come, if not from the job itself?

What kind of configuration change could exist but in an "unapplied" state, that this option would apply?

Should it read 

> applying build discarder rules even when no new builds are run

or is it really talking about job configuration?

daniel-beck replied:
> or is it really talking about job configuration?

Well, build discarder rules are a part of the job configuration. So in a sense, yes.

Currently, you'd need to run a new build to apply a changed build discarder configuration, since it's only consulted after a build finishes. So without new builds, you can change this option and it doesn't have any consequences. For an obsolete/archived job, it would never actually delete builds/artifacts.

Perhaps

> …applying changes to a job's build discarder configuration even when…

?

awittha commented on core/src/main/java/jenkins/model/JobGlobalBuildDiscarderStrategy.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    public void apply(Job<?, ?> job) throws IOException, InterruptedException {
+        job.logRotate();
```
`Job` is a non-final, public, abstract class, and `logRotate()` is a non-final, public method on `Job` - and neither are annotated with `@NoExternalUse`.

It is not guaranteed that `job.logRotate()` will actually use the build discarder that was checked in `isApplicable(...)`. This may cause the documentation around this step to fail to align with what actually happens.

Is there a reason to prefer `job.logRotate()` over `job.getBuildDiscarder().perform( job )`?

daniel-beck replied:
Good question. I think the expectation is that `#logRotate()` not be overridden, but it's not enforced, true.

I am unsure which is the more correct option.

`#logRotate()` is called by `Run#execute(RunExecution)`, so an argument can be made that this is the expected way to rotate logs, and if the job overrides it in a way that differs from what `getBuildDiscarder().perform(…)` does, `#logRotate()` should be take precedence.

OTOH, the way this feature is explained on the UI would indicate we want `getBuildDiscarder().perform(…)` instead. Obviously, much of that is by necessity.

awittha commented on core/src/main/java/jenkins/model/BackgroundGlobalBuildDiscarder.java:
```diff
@@ -0,0 +1,80 @@
+    private static final Logger LOGGER = Logger.getLogger(BackgroundGlobalBuildDiscarder.class.getName());
+
+    public BackgroundGlobalBuildDiscarder() {
+        super("Periodic background build discarder"); // TODO i18n
```
Is the `TODO` required for merge?

daniel-beck replied:
I don't think so. I'm not sure we're even localizing these.

timja commented:
> Should we really limit each descriptor to one occurrence?

Would be good to allow multiple so that I.e a regex based extension could have different rules for master branch and other branches / prs

daniel-beck commented:
> Would be good to allow multiple so that I.e a regex based extension could have different rules for master branch and other branches / prs

This is still possible, just requires more complex individual discarders (as they'd need to support N patterns/rules instead of 1).

Forgot to mention a downside: Especially the default (unconfigurable) jobBuildDiscarder could be added multiple times, possibly leading to user confusion; it seems to me it also potentially be more difficult to write build discarders that have rules that depend on a larger context (e.g. builds across jobs).

timja commented:
hmm I guess that would be fine

daniel-beck commented:
> hmm I guess that would be fine

Could you clarify which behavior is preferable in your opinion?

timja commented:
More complex individual discarders is fine, 
so that you don’t get multiples of discarders that shouldn’t be added multiple times

jvz commented on core/src/main/resources/jenkins/model/GlobalBuildDiscarderConfiguration/config.groovy:
```diff
@@ -0,0 +1,33 @@
+
+f = namespace(lib.FormTagLib)
+
+f.section(title: _("Global Build Discarders")) {
```
Nit: does this view really need to be Groovy? Seems fairly trivial to convert to Jelly.

jvz reviewed (approved):
Looks good overall. Would certainly like to see this feature in production where disk space errors like to hang out.

timja commented:
@daniel-beck are the questions in the PR description still open / needing feedback?
```
Open Questions (feedback welcome):

Should the "Job" Global Build Discarder be implemented differently, without a UI option? If we actually consider it to be a bug fix (see above and below), why would there be a UI that boils down to "[x] fix the bug"?
Should we really limit each descriptor to one occurrence? This should make it more difficult to implement more flexible strategies, e.g. the proposed "regex match" one would need its own list of (regex/build discarder) pairs, rather than just one of each as a property, and then have N instances of the extension in the top level list…
```
There's 4 approvals on this, just checking if everything is in order to move forward on it

daniel-beck commented:
@timja 

> are the questions in the PR description still open / needing feedback?

I have received no responses to the first question. I'm fairly OK with the way it works, just flagging it for attention if someone cares. It doesn't look like it so far.

I have only received one response to the second question (from you), and it was approving of the current implementation after a short conversation. FWIW this can always be opened up further, but restricting would be difficult.

There's one unaddressed feedback of the code catching an `InterruptedException` when it shouldn't, but have held off addressing that until I got more feedback to limit iteration count. It doesn't look like that will be forthcoming, so perhaps it's time to address that so this can be merged.

daniel-beck committed:
Do not catch InterruptedException

timja commented:
There appears to be no outstanding issues and 4 approvals, looks like it's time to get this in.

After a 24 hour timeout for last minute feedback, this can be merged.
